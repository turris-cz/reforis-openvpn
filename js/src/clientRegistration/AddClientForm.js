/*
 * Copyright (C) 2020-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";

import {
    TextInput,
    Button,
    useAPIPost,
    useForm,
    useAlert,
    undefinedIfEmpty,
    API_STATE,
} from "foris";
import PropTypes from "prop-types";

import API_URLs from "API";

AddClientForm.propTypes = {
    generating: PropTypes.bool.isRequired,
    setGenerating: PropTypes.func.isRequired,
};

export default function AddClientForm({ generating, setGenerating }) {
    const [setAlert] = useAlert();

    const [postClientsResponse, postClients] = useAPIPost(API_URLs.clients);
    useEffect(() => {
        if (postClientsResponse.state === API_STATE.ERROR) {
            setAlert(postClientsResponse.data);
            setGenerating(false);
        }
    }, [postClientsResponse, setAlert, setGenerating]);

    const [formState, formChangeHandler, reloadForm] = useForm(
        addClientFormValidator
    );
    const formData = formState.data;
    const formErrors = formState.errors || {};
    useEffect(() => {
        reloadForm({ name: "" });
    }, [reloadForm]);

    function handleSubmit(event) {
        event.preventDefault();
        postClients({ data: formData });
        setGenerating(true);
    }

    if (!formData) {
        return null;
    }

    const addButtonDisabled = undefinedIfEmpty(formErrors) || generating;
    return (
        <>
            <h2>{_("Add New Client")}</h2>
            <form onSubmit={handleSubmit}>
                <TextInput
                    label={_("Client name")}
                    helpText={_(
                        "Shorter than 64 characters. Only alphanumeric characters, dots, dashes and underscores."
                    )}
                    value={formData.name}
                    error={formErrors.name}
                    onChange={formChangeHandler((value) => ({
                        name: { $set: value },
                    }))}
                />
                <div className="text-right">
                    <Button
                        type="submit"
                        forisFormSize
                        disabled={addButtonDisabled}
                    >
                        {_("Add")}
                    </Button>
                </div>
            </form>
        </>
    );
}

function addClientFormValidator(formData) {
    const { name } = formData;
    if (!name) {
        return { name: _("Name cannot be empty.") };
    }
    if (name.length > 64) {
        return { name: _("Name is too long.") };
    }
    if (!/^[a-zA-Z0-9.\-_]+$/.test(name)) {
        return { name: _("Name contains invalid characters.") };
    }
    return undefined;
}
