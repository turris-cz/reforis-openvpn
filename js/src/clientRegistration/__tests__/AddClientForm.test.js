/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render, fireEvent, wait, getByText, getByLabelText,
} from "customTestRender";
import mockAxios from "jest-mock-axios";

import { AlertContext } from "foris";

import AddClientForm from "../AddClientForm";

function getFormElements(componentContainer) {
    return {
        nameInput: getByLabelText(componentContainer, "Client name"),
        submitButton: getByText(componentContainer, "Add"),
    };
}

describe("<AddClientForm />", () => {
    const setAlert = jest.fn();
    let componentContainer;

    beforeEach(() => {
        const { container } = render(
            <AlertContext.Provider value={setAlert}>
                <AddClientForm />
            </AlertContext.Provider>,
        );
        componentContainer = container;
    });

    it("should be rendered", () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("should validate form", () => {
        const { nameInput, submitButton } = getFormElements(componentContainer);

        // Empty name
        fireEvent.change(nameInput, { target: { value: "" } });
        expect(getByText(componentContainer, "Name cannot be empty")).toBeDefined();
        expect(submitButton.disabled).toBe(true);

        // Name too long
        fireEvent.change(nameInput, { target: { value: "q".repeat(65) } });
        expect(getByText(componentContainer, "Name is too long")).toBeDefined();
        expect(submitButton.disabled).toBe(true);

        // Invalid characters
        fireEvent.change(nameInput, { target: { value: "!@#$%" } });
        expect(getByText(componentContainer, "Name contains invalid characters")).toBeDefined();
        expect(submitButton.disabled).toBe(true);
    });

    it("should add client", () => {
        const { nameInput, submitButton } = getFormElements(componentContainer);

        expect(submitButton.disabled).toBe(true);
        fireEvent.change(nameInput, { target: { value: "propername" } });
        expect(nameInput.value).toBe("propername");
        expect(submitButton.disabled).toBe(false);

        fireEvent.click(submitButton);
        expect(mockAxios.post).toBeCalledWith(
            "/reforis/openvpn/api/clients",
            { name: "propername" },
            expect.anything(),
        );
    });

    it("should handle API error", async () => {
        const { nameInput, submitButton } = getFormElements(componentContainer);

        fireEvent.change(nameInput, { target: { value: "propername" } });
        fireEvent.click(submitButton);

        const errorMessage = "API didn't handle this well";
        mockAxios.mockError(
            { response: { data: errorMessage, headers: { "content-type": "application/json" } } },
        );
        await wait(() => {
            expect(setAlert).toHaveBeenCalledWith(errorMessage);
        });
    });
});
