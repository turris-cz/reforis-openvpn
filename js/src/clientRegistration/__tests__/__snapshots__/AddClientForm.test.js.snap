// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddClientForm /> should be rendered 1`] = `
<div>
  <h2>
    Add New Client
  </h2>
  <form>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="1"
      >
        Client name
      </label>
      <div
        class="input-group"
      >
        <input
          class="form-control is-invalid"
          id="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="invalid-feedback"
      >
        Name cannot be empty.
      </div>
      <div
        class="form-text"
      >
        <small>
          Shorter than 64 characters. Only alphanumeric characters, dots, dashes and underscores.
        </small>
      </div>
    </div>
    <div
      class="text-end"
    >
      <button
        class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
        disabled=""
        type="submit"
      >
        <span>
          Add
        </span>
      </button>
    </div>
  </form>
</div>
`;
